x-pack/plugins/stack_connectors/server/connector_types/gen_ai/lib/utils.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { AxiosRequestHeaders, AxiosResponse, ResponseType } from 'axios';
+import { OpenAiProviderType } from '../../../../common/gen_ai/constants';
+import {
+  sanitizeRequest as openAiSanitizeRequest,
+  getRequestWithStreamOption as openAiGetRequestWithStreamOption,
+} from './openai_utils';
+import {
+  sanitizeRequest as azureAiSanitizeRequest,
+  getRequestWithStreamOption as azureAiGetRequestWithStreamOption,
+} from './azure_openai_utils';
+
+export const sanitizeRequest = (provider: string, url: string, body: string): string => {
+  switch (provider) {
+    case OpenAiProviderType.OpenAi:
+      return openAiSanitizeRequest(url, body);
+    case OpenAiProviderType.AzureAi:
+      return azureAiSanitizeRequest(url, body);
+    default:
+      return body;
+  }
+};
+
+export const getRequestWithStreamOption = (
+  provider: string,
+  url: string,
+  body: string,
+  stream: boolean
+): string => {
+  switch (provider) {
+    case OpenAiProviderType.OpenAi:
+      return openAiGetRequestWithStreamOption(url, body, stream);
+    case OpenAiProviderType.AzureAi:
+      return azureAiGetRequestWithStreamOption(url, body, stream);
+    default:
+      return body;
+  }
+};
+
+export const getAxiosOptions = (
+  provider: string,
+  apiKey: string,
+  stream: boolean
+): { headers: AxiosRequestHeaders; responseType?: ResponseType } => {
+  const responseType = stream ? { responseType: 'stream' as ResponseType } : {};
+  switch (provider) {
+    case OpenAiProviderType.OpenAi:
+      return {
+        headers: { Authorization: `Bearer ${apiKey}`, ['content-type']: 'application/json' },
+        ...responseType,
+      };
+    case OpenAiProviderType.AzureAi:
+      return {
+        headers: { ['api-key']: apiKey, ['content-type']: 'application/json' },
+        ...responseType,
+      };
+    default:
+      return { headers: {} };
+  }
+};
+
+export const pipeStreamingResponse = <R>(response: AxiosResponse<R>) => {
```

Comment:
I was too lazy to strip the headers of the OpenAI request, that's probably something we should do too?

Reply:
Sorry, do you mean strip the headers of the response so the only header on the response is the `Content-Type: dont-compress-this`? We don't strip response headers from other axios responses, are there specific headers you're concerned about?

Reply:
Are there other cases where the response headers are propagated? I'm not sure if there are sensitive headers, but for instance, I had to work around a 401 being returned from OpenAI that caused the session to be invalidated, requiring the user to log in again.

Reply:
The connector framework handles error responses and wraps them in a custom structure that is returned by the actions client. I've updated the example route to check for the error status https://github.com/elastic/kibana/pull/161676/commits/b6338431082a14d61c243d9dda685c2a98fde2e8

Reply:
In general, does seem like we'll need to somehow "wrap" this.  Anything returned from this API will appear as if Kibana itself returned it.  There could be security leakage of headers and such.  

Someone on the Kibana side should probably be reading the entire stream and sending the bits down manually.  At first I thought this was perhaps best done in a solution plugin, but that will have the same issue - the raw response from the service is being spit out of Kibana as if it's a Kibana response, which it's not.

So, feels like it should go here - in the rule-type specific code.  I'm guessing it will need some tweakage over time.  Like, it should be aware of when the client request closes, so it can close the back-end stream.  Maybe it has it's own timeouts, or default timeouts need to be extended ...

Also not clear to me how this is currently working, as it's setting a header in `response`, but then sending `response.data` - does that somehow include the headers?  Not sure what `response.data` even is here TBH :-)

Reply:
For streaming, the `response.data` is of type `IncomingMessage` which extends a readable stream: https://nodejs.org/api/http.html#class-httpincomingmessage and does contain the request & response headers. I will look into sanitizing this

Reply:
@pmuellr I looked into this some more. As mentioned above, when the request returns an error, the connectors framework already handles this and wraps the error so the situation with the 401 response headers being returned should not occur. When the request returns successfully, it returns an `http.IncomingMessage` which does contain the original client request with headers. However, when the `IncomingMessage` is returned via the server route, it is wrapped inside the `KibanaResponseFactory`. After pinging the core team, it seems that the Kibana response factory accepts a `Streamable` (which `IncomingMessage` extends) which only takes the body of the stream. Testing this out with POSTMAN, I see only the data body being returned and the response headers are the `Content-Type` we set and the ones set by the `KibanaResponseFactory`.

Reply:
Out of interest, can we still handle the original status code in some way when using this with the actionsClient?

Reply:
Do you mean on error? Right now it looks like the status code will be embedded in the error message but we could open a followup issue to return just the status code.

Reply:
No that's fine I think - just making sure we can eg still tell the user that authentication failed or they are being rate limited